### ISSUE TITLE
False positive  `used-before-assignment` when usage and assignment guarded by the same test in different statements

### ISSUE THREAD

**Issue by tvandeveenzepp** [pylint-dev/pylint#8167]
### Bug description

The `used-before-assignment` check reports false positive for using the same variable inside multiple `if __name__ == '__main__'` statements.

```
# pylint: disable=missing-docstring

if __name__ == '__main__':
    X = 1

if __name__ == '__main__':
    print(X)
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:7:10: E0601: Using variable 'X' before assignment (used-before-assignment)

-----------------------------------
Your code has been rated at 0.00/10
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 2.16.0
astroid 2.14.1
Python 3.7.4 (tags/v3.7.4:e09359112e, Jul  8 2019, 20:34:20) [MSC v.1916 64 bit (AMD64)]
```

**Comment by clavedeluna** [comment]
It would appear `if __name__ == '__main__':` is causing different behavior because if you do:
```
# pylint: disable=missing-docstring

if True:
    X = 1

if True:
    print(X)
```

the msg is not emitted. My hunch is this is happening because pylint isn't loading it as a script which is what `if __name__ == '__main__':` requires.

**Comment by jacobtylerwalls** [comment]
Likely met the "always false" test in #6677.

**Comment by jacobtylerwalls** [comment]
We may need to check if the variable is guarded again by the same(ish) test.

**Comment by saranti** [comment]
This issue wasn't present in v2.15.10. I'll take this up if no one objects

**Comment by jacobtylerwalls** [comment]
This also affects `if TYPE_CHECKING:` e.g.:

```python
if TYPE_CHECKING:
    import math

if TYPE_CHECKING:
    assert math
```

This is because `TYPE_CHECKING`, like `__name__ == '__main__'`, is false at runtime except under special circumstances.
